### ISSUE TITLE
cd to the working directory of another elvish session

### ISSUE THREAD

**Issue by xiaq** [elves/elvish#190]
When I cd to a certain directory, I often want to cd to the same directory in another session. It is nice to have the working directories of other sessions marked specially in the location mode.

For instance, we can put a `*` in front of the working directories of active sessions, and having `*` in the filter restricts the candidates to those directories.

**Comment by hanche** [comment]
Stray thought on this: Since all sessions are talking to the same daemon process anyway, why not use this fact to implement a sort of message bus allowing the shells to communicate between themselves on a more general level? The current working directory is only one of many possible pieces of data that one might want to share between sessions, after all.

**Comment by krader1961** [comment]
> It is nice to have the working directories of other sessions marked specially in the location mode.

I don't see how this is useful other than very narrow scenarios. If I have ten elvish sessions on a particular system, all of which have a different CWD, how useful is it for the location mode to mark all ten directories as being a CWD in another session? What happens to the location mode display if I terminate one of those sessions (e.g., via [ctrl-D])? Why not just copy/paste the output of `pwd`?

This seems to me a lot like the use case for merging global command history with local history via the `edit:history:fast-forward` command so that a session can "see" the commands recently run in other sessions. In this case the global history would contain the most recent directory change in any session. Merging global history into local history would expose the global "most recent cd destination" as the first entry shown by location mode so that simply pressing [enter] would select it. Merging of the global "most recent cd" could be done implicitly on every invocation of location mode rather than requiring an explicit `edit:history:fast-forward`.

Whether implicitly merging the global CWD is a good idea is not obvious. Especially since it is not obvious that we should, or even can, automatically synchronize with the global CWD state without also syncing command history. At least without using a message bus/queue as proposed by @hanche. The problem with that solution is that whether or not we use OS specific features (e.g., `man 7 mq_overview` on Linux) it requires a lot of additional baggage.

**Comment by hanche** [comment]
Just to be clear on this, I am not really in favour of the message bus proposal. As @krader1961 says, it does require quite a lot of baggage, though likely less than more general solutions intended for more general purposes. But I think I do favour it over sharing CWD information, being a more general mechanism but not necessarily much more complex. (You'd have to be careful to keep it as simple as possible.)

**Comment by krader1961** [comment]
LIke @hanche I am ambivalent about including support for an elvish message bus. On the one hand it might enable interesting features. On the other hand it adds a lot of baggage that can't be justified solely to support sessions sharing CWD changes.

**Important:** I would like to see the most recent CWD change by any other elvish session appear as the first entry in the location mode of a particular elvish session (regardless of whether the change is done using location mode or an explicit `cd` command). That would make it easier to do an explicit `cd /a/long/path` in one session and quickly switch to it in a different session without resorting to copy/paste or explicitly syncing command history via `edit:history:fast-forward`. That's something I do with some regularity. What I don't think is workable, or particularly useful, is the idea of making the CWD of every other elvish session visible in the location mode of a particular elvish session.

P.S., Note that this feature requires converting relative paths (e.g., `cd .`) into absolute paths when exposed to other elvish sessions.

**Comment by xiaq** [comment]
> I would like to see the most recent CWD change by any other elvish session appear as the first entry in the location mode of a particular elvish session

Indeed, considering *recently CWD changes* is much simpler than considering CWD of active sessions.

Instead of treating the most recent CWD change specifically, there is a more general solution: in location mode, support ordering the directory list purely by recency. This can be part of a general mechanism of supporting custom ordering in listings - so not just location mode, but also history listing mode, completion mode, and the individual columns of navigation mode.

**Comment by krader1961** [comment]
> ... in location mode, support ordering the directory list purely by recency.

That solution works for me assuming I don't need to run `edit:history:fast-forward`. That is, I want CWD changes to be immediately visible in the location mode of other shells but do not want commands run in a shell to be immediately visible in the other shells.
